### ISSUE TITLE
Print CSS Guide

### ISSUE THREAD

**Issue by cfarm** [cfpb/design-manual#281]
I am working on print styles for Owning a Home. There are a few print styles in the works for the CF.Gov Refresh. I find myself making some design decisions on what elements to change for print, and there are no guidelines on print CSS in the Design Manual. 

For now, I'm taking direction from the [HTML5 Boilerplate print styles](https://github.com/h5bp/html5-boilerplate/blob/master/src/css/main.css) (also used by Bootstrap).

Other resources on this:
- [Owning a Home Print styles, in progress](https://github.cfpb.gov/OAH/OAH-notes/issues/796)
- ["What's the level of print-friendliness of current site?"](https://github.cfpb.gov/OAH/OAH-notes/issues/775)
- [cfgov-refresh print styles approach](https://github.com/cfpb/cfgov-refresh/issues/248)
- [docs for cf-gov refresh print styles, in progress](http://cfpb.github.io/cfgov-refresh/docs/print/)

**Comment by jenn-franklin** [comment]
Starting to look further at this. Creating a list of thoughts so far from the OaH team from github issues and conversations: 

<b>Color</b>
- All page elements will print in black and white. Should this apply to the header (including cfpb logo) and footer?
- Background colors in areas such as sidebars and header bars will drop off. 
- Button colors will drop out. Button text will remain with gray border around it. 

<b>Images</b>
- Images will not print. This applies to header graphics, illustrations, photos, isocons. Does not apply to minicons. Does not apply to content-driven images such as graphs like on the rate checker page. Do social media icons print in the footer?

<b>Links</b>
- Href attribute will appear to the side and/or underneath link text. Will be underlined and in parentheses. Should this path be relative or include "consumerfinance.gov?" Have been looking at other sites for reference. So far have found that gov.uk includes full path. Mint and Warby Parker use relative paths. 
- Href attribute should appear for header and footer links, including social media?
- Href attribute should appear with any links embedded in copy.
- Output URLs for media like youtube, videos, share? Or hide them?

<b>Lines</b>
- Lines separating content and as sidebar headers will print.

<b>Expandables</b>
- Should expandables on a page collapse?
- Should the minicon/text of an expandable button appear or disappear or appear in a different style when printed?

<b>Typography</b>
- Will stay in web type heirarchy and not convert to print type?

**Comment by Dnpizarro** [comment]
@huetingj will work on a mock up for the print styles.

**Comment by jenn-franklin** [comment]
To explore different aspects of this, I've used http://refresh.demo.cfpb.gov/doing-business-with-us/

![screen shot 2015-04-08 at 10 11 24 am](https://cloud.githubusercontent.com/assets/10562530/7047022/dcb1f576-ddd7-11e4-9253-954bb92a8966.png)

**Comment by jenn-franklin** [comment]
<b>1. All elements on a page will become grayscale when printing.<b/>

![screen shot 2015-04-08 at 10 15 10 am](https://cloud.githubusercontent.com/assets/10562530/7047075/3381063a-ddd8-11e4-98ea-fcae4168d5f5.png)

This includes the cfpb logo in the header and footer as well as the lines, which I've changed to 50% black. 

![screen shot 2015-04-08 at 10 17 14 am](https://cloud.githubusercontent.com/assets/10562530/7047110/7886402e-ddd8-11e4-88ac-80799d2e4cf9.png)

![screen shot 2015-04-08 at 10 17 22 am](https://cloud.githubusercontent.com/assets/10562530/7047115/7d69e7a8-ddd8-11e4-89b0-da507688071f.png)

**Comment by jenn-franklin** [comment]
<b> 2. Background colors will drop out.<b/>

So, in this case, the gray box:

![screen shot 2015-04-08 at 10 20 35 am](https://cloud.githubusercontent.com/assets/10562530/7047178/ed447d7c-ddd8-11e4-8887-5bd261638f3f.png)

Becomes this: 

![screen shot 2015-04-08 at 10 21 03 am](https://cloud.githubusercontent.com/assets/10562530/7047191/ffe59d62-ddd8-11e4-92cd-9b1e99e45143.png)

**Comment by jenn-franklin** [comment]
<b>3. Minicons will print (and become black).<b/>

![screen shot 2015-04-08 at 10 22 02 am](https://cloud.githubusercontent.com/assets/10562530/7047211/29a71cd4-ddd9-11e4-8cc3-b1de27457667.png)

![screen shot 2015-04-08 at 10 22 17 am](https://cloud.githubusercontent.com/assets/10562530/7047213/2cd3bc8c-ddd9-11e4-9096-d63ed44ac79d.png)

**Comment by jenn-franklin** [comment]
<b>4. Link paths will appear.<b/>

A.) For pages within consumerfinance.gov, should these paths be relative?

![screen shot 2015-04-08 at 10 52 12 am](https://cloud.githubusercontent.com/assets/10562530/7047860/5ab2abaa-dddd-11e4-9063-7955d87fd530.png)

B.) Styling: Link text becomes Avenir Next Regular, black and with a solid underline. Link path appears directly next to/under link text in parentheses and with a solid underline. 

![screen shot 2015-04-08 at 10 30 44 am](https://cloud.githubusercontent.com/assets/10562530/7047411/5819e730-ddda-11e4-94ba-c6d01bf8453a.png)

Another possibility is to make the link text demi bold to pull it out a little from the path:

![screen shot 2015-04-08 at 10 32 48 am](https://cloud.githubusercontent.com/assets/10562530/7047451/a2aa6734-ddda-11e4-90cb-7b79b85bc8cf.png)

C.) Links in the the side nav, header and footer: We need to consider whether or not to make the paths appear and if they should be relative. 

Header example:
![screen shot 2015-04-08 at 10 37 34 am](https://cloud.githubusercontent.com/assets/10562530/7047599/8c704e4c-dddb-11e4-9096-0d16ab62ea5b.png)

Side nav:
![screen shot 2015-04-08 at 10 57 25 am](https://cloud.githubusercontent.com/assets/10562530/7047993/147c90c8-ddde-11e4-996e-f33f7c6f176f.png)

Same consideration regarding social media icons and whether or not the links are shown relatively to those sites:
![screen shot 2015-04-08 at 10 58 35 am](https://cloud.githubusercontent.com/assets/10562530/7048032/3efc4e42-ddde-11e4-83d1-d4fe23738a22.png)

D.) Example of pulling an email address out:
![screen shot 2015-04-08 at 10 59 12 am](https://cloud.githubusercontent.com/assets/10562530/7048036/5053af82-ddde-11e4-839e-811dc60a1122.png)

**Comment by jenn-franklin** [comment]
<b>6. All gray separator lines will print.<b/>

![screen shot 2015-04-08 at 11 06 31 am](https://cloud.githubusercontent.com/assets/10562530/7048181/57df8946-dddf-11e4-98de-d4b3919190f4.png)

**Comment by Scotchester** [comment]
I think we should ask the question of whether things like navigation and social media links should be printed at all. Most likely, that stuff is useless to the person who wants to have a printed version of the page.

**Comment by jenn-franklin** [comment]
That's a good thought. Perhaps the print header/footer just include our logo, phone number and email--- or some other very basic identifying elements.

**Comment by jenn-franklin** [comment]
Page view with basic header and footer info and no navigation. Also no side nav, which, if we drop, we might want to contemplate if that changes anything with the spacing/margins of the printed page. 

![screen shot 2015-04-08 at 11 24 48 am](https://cloud.githubusercontent.com/assets/10562530/7048677/7531c57e-dde2-11e4-8668-bc187734551a.png)

**Comment by schaferjh** [comment]
Hmmm, dropping the side nav may drop important contextual information. Not in _all_ cases but definitely in some. I would be wary of dropping that all together. (Though taking the link paths out for that side nav makes sense.)

I like the look of dropping the header and footer though. Interesting exploration!

**Comment by Scotchester** [comment]
@schaferjh I did have that thought, too.

I haven't followed your breadcrumb discussions super closely, but would that proposal provide enough of that contextual information, if those were printed (without the link paths)?

**Comment by schaferjh** [comment]
No, the breadcrumb only shows the path to the page, not the sibling or children pages. So in the example Jen is using above, the breadcrumb alone wouldn't inform people that there is further information available around future requests, active requests and small businesses. 

Depending on why someone has  printed out the page, this could be important info. (Again, not on every page but on enough of them that removing it by default seems like a step too far.)

**Comment by schaferjh** [comment]
Discussion during 4/23 design manual working session with @huetingj and @Dnpizarro:
- We like displaying the link information in body content (email or full URL), but leave any navigation/menu links as straight text, no relative links showing.
- We like the idea of providing some contrast between the link text and displayed URL, per @huetingj's example above where she made the link text demi bold. (She's going to explore a little more on this.)
- Keep the header and footer intact and without any relative links showing.

Woot!

**Comment by jenn-franklin** [comment]
Here are a few ways the links could be styled. 
1. Link text is demi-bold
   ![demibold](https://cloud.githubusercontent.com/assets/10562530/7454133/d77b9758-f23e-11e4-80b1-583013607e67.png)
2. Link text is medium
   ![medium](https://cloud.githubusercontent.com/assets/10562530/7454139/ecda03a0-f23e-11e4-8391-ae7b653c90bf.png)
3. Link text is regular and URL is dark grey
   ![regulargray](https://cloud.githubusercontent.com/assets/10562530/7454144/fd20b952-f23e-11e4-94b9-ef3e6241174f.png)
4. Link text is medium and URL is dark grey
   ![mediumgray](https://cloud.githubusercontent.com/assets/10562530/7454147/099e4f50-f23f-11e4-921a-9c0f9b058c31.png)
5. Link text is demi-bold and URL is dark grey
   ![demiboldgray](https://cloud.githubusercontent.com/assets/10562530/7454161/34832c04-f23f-11e4-943f-7ee87531bb28.png)

**Comment by Dnpizarro** [comment]
@huetingj it helps to see the various options for the links. I like your solution for the url, I feel the dark grey gives enough of that distinction that we were looking for. My tendency is to lean towards option 3 since it stays inline with our our link standard. However, I also like option 4. @schaferjh how are you feeling about the new options?

**Comment by schaferjh** [comment]
Nice. I agree that I like using the dark gray for the URL text. Both 3 and 4 work for me, so if 3 is more in line with our current standard, I say let's go in that direction.

**Comment by sonnakim** [comment]
Nicholas Johnson (sorry, can't find his github handle!) found a government URL shortener: https://go.usa.gov. 

Shortened URL example:
before: http://build.consumerfinance.gov/owning-a-home/loan-comparison
after: http://go.usa.gov/3K6Ue 

Authentication occurs by signing up with a government email address, and the URLs never expire.  

Any considerations or issues before I use these for my printable PDF needs?

**Comment by schaferjh** [comment]
A few thoughts:
*For the printed experience, I'm worried that if all the links display as "http://go.usa.gov/..." it will look like some kind of gibberish the printer/computer did to the website on print. Esp. for less-sophisticated users, it removes one of the benefits of showing the URL -- trust and confirmation.
*For the actual web experience, we are going to be implementing a privacy notice interstitial for non-gov third party links in the refresh. If the links are all entered as shortened .gov versions, then we won't know when to trigger the privacy notice and when not to. If the shortened versions have to be entered separately from the actual links as a "print only link", that strikes me as something of a burden and prone to human error.

I know the shortened links are super attractive, but I'm not sure they're really the right answer for the print style.

**Comment by Scotchester** [comment]
I agree with everything @schaferjh said, but I will point out that go.usa.gov can only shorten .gov URLs, so there is no uncertainty over whether something is a .gov link or not.

**Comment by schaferjh** [comment]
Ah ha -- thanks @Scotchester for that clarification!

**Comment by benguhin** [comment]
Those are really good points @schaferjh and @Scotchester – it seems like it's best to not shorten the URLs. 

If the user doesn't want to type in a super-long URL, they can always go back to the web page they printed and click on the link from there. Including the real URL in the printout provides additional context for users to learn where they would have been taken (e.g., "Oh, that link goes to the FTC website")

**Comment by benguhin** [comment]
I want to challenge the assumption that the print styles should be grayscale. Some people use the Print function to save the page as a PDF, and other people will prefer to print in color. I think the folks who prefer to print in black and white can be expected to opt-in by choosing black and white in their print settings.

For the example in this thread, that mostly impacts our logo and the links on the page, but it'll be more impactful for pages that include photos or illustrations.

**Comment by KimberlyMunoz** [comment]
+1 to @benguhin's comment on grayscale on images.

I think "All images will print in grayscale" is an assumption we can make about how a good chunk of folks will end up printing these documents (and it's really great that the images in this discussion are designing around that assumption), but it's not something we should enforce as a rule.

**Comment by jenn-franklin** [comment]
Pulling these thoughts into one view:

![screen shot 2015-05-29 at 11 00 19 am](https://cloud.githubusercontent.com/assets/10562530/7885480/084e8964-05f2-11e5-9595-b65572f7c39f.png)
- All navigation will remain when printed and not include URLs.
- Links within page content will include full URLs. A different font weight will be used to differentiate the link text from the URL. Above, I'm using demi bold for link text and regular for URLs.
- Background colors will drop out. This seems nonessential but keeping it here for further consideration.
- All elements will print/export to pdf in color. It will be the user's responsibility to select BW in his print settings if preferring no color. 
- All images will print. 
- Expandables a user has collapsed will remain collapsed when printed. Ones a user has not collapsed will remain rolled up on print.

**Comment by stephanieosan** [comment]
What about illustrations or heros? Will those stay or drop out like the background colors?

**Comment by jenn-franklin** [comment]
I'm leaning toward keeping them. When I look at sections such as "About the Bureau" on beta, I realize that photos such as the director's headshot provide helpful context and should print. Many of our illustrations are supplemental but are still helpful in creating a tone for a page. With headers, removing the background image but keeping the text seems like it would get funky space-wise and might look like a mistake, especially if someone is only exporting to pdf.

**Comment by stephanieosan** [comment]
:+1: That makes sense to me.

**Comment by jenn-franklin** [comment]
It was decided today in our design manual meeting that expandables a user has collapsed will remain collapsed when printed. Ones a user has not collapsed will remain rolled up on print.

**Comment by jenn-franklin** [comment]
Hi, @JenniferHoran! We're putting together design guidelines for how our web pages should print. Here is a sample page reflecting our decisions:

![084e8964-05f2-11e5-9595-b65572f7c39f](https://cloud.githubusercontent.com/assets/10562530/7947127/1679d0d4-0949-11e5-81a4-5419b8e613a2.png)

Can you please review these points and make sure we're not overlooking anything important in terms of accessibility?
- [ ] All navigation elements in the header, left side area and footer will remain when printed. URLs will not appear for links in these sections.
- [ ] Full URLs will appear for links within a page's main content. Links will remain Pacific blue. Avenir demi-bold weight will be used for link text and regular Avenir will be used for URLs. The difference in weight is intended to help users easily separate the link text from the path. 
- [ ] Pages will print in full color.
- [ ] All images will print.
- [ ] Expandables a user has collapsed will remain collapsed when printed. Expandables a user has not collapsed will remain rolled up when printed. Here is a view of expandables for reference:

![screen shot 2015-06-02 at 5 11 47 pm](https://cloud.githubusercontent.com/assets/10562530/7947362/9f32853c-094a-11e5-962e-23054a6bf64c.png)

Thanks for your thoughts!

**Comment by JenniferHoran** [comment]
Yes, I'd love the opportunity to go over this with you but I'm out of pocket this afternoon offsite and on leave 5-10 June, back on the 11th. I can take a look tomorrow, but his might have to wait until I return? 
Thank you!

**Comment by jenn-franklin** [comment]
Ok, no problem and sounds good! Thank you!

**Comment by JenniferHoran** [comment]
Hi! Still  reading through all of the comments in this thread, but one best practice for CSS to print, for 508 especially, is EEOC http://www.eeoc.gov/employees/afterfiling.cfm

**Comment by sonnakim** [comment]
@huetingj Just to confirm, are we still planning to <a href="https://github.com/cfpb/design-manual/issues/281#issuecomment-90930923"> drop background colors out</a>?

**Comment by jenn-franklin** [comment]
@sonnakim many browsers will automatically drop off background colors.

**Comment by jenn-franklin** [comment]
Hi @JenniferHoran! Any more thoughts on this? We're going to wrap up these basic guidelines soon and start a new issue addressing more complicated print decisions as they arise. Thanks!

**Comment by jenn-franklin** [comment]
Sorry for the drop off. Picking this back up with gusto.  

Here are the proposed guidelines. They are basic. As the need for custom print styles comes up, teams should share their issues and solutions to identify overlap/broader needs for standardization. 
- All navigation elements in the header, left side area and footer will remain when printed. URLs will not appear for links in these areas.
- Full URLs will appear for links within a page's main content. Links will remain in blue. Avenir demi-bold weight will be used for link text and regular Avenir will be used for URLs. The difference in weight is intended to help users easily separate the link text from the path. 
- Pages will print in full color, and all images will print. The reasoning behind this is to keep color and images when a person prints the page as a pdf. I think this is particularly important as we push to move away from pdfs in general and toward basic web pages. A user can choose B&W mode if desired when printing on paper.
- Expandables a user has collapsed will remain collapsed when printed. Expandables a user has not collapsed will remain rolled up when printed. 

I don't really think we need a visual example when we put this into a design manual page. But here's this example again for reference particularly in how the text weight looks for links. (Although a note that the links in the example below are in Pacific blue, and I believe the plan is to change our standard to dark Pacific blue.)

![084e8964-05f2-11e5-9595-b65572f7c39f](https://cloud.githubusercontent.com/assets/10562530/7947127/1679d0d4-0949-11e5-81a4-5419b8e613a2.png)

V1 folks, @schaferjh @duelj @ajbush @jimmynotjim Just to be sure: Have you developed any print styles in CF since this summer? Any objections to the above guidelines?

@JenniferHoran any feedback?

@benguhin @nataliafitzgerald Can you please review and approve?

Will need a FEWD to build this out. If my team is unable to take it on, I will reach out to others and/or bring it up during the next design manual monthly meeting. 

I will start putting this into a design manual page.

**Comment by jimmynotjim** [comment]
We haven't done anything unfortunately, but I'm more than happy to provide input and reviews.

**Comment by schaferjh** [comment]
Thanks for picking this back up @huetingj. Looks good to me -- I don't see anything here that's contrary to or an issue for any of our work. 

I agree re: visual example not being needed for the DM page, but on the off chance you do need one, let us know so we can provide an updated footer/header (or you might want to use the current footer/header depending on when this goes live.)

**Comment by benguhin** [comment]
@huetingj this looks great. I have no objections to those guidelines!

**Comment by jenn-franklin** [comment]
@nataliafitzgerald Can you please approve? Thanks!

**Comment by marteki** [comment]
Printing the Dynamic Disclosures tool is a part of the scope of the project, and I'll be working on the styles for that during this sprint (three weeks, today through Feb 9th).

So, if you need a FEWDish type person to work on it, @huetingj, I'm volunteering.

**Comment by jenn-franklin** [comment]
Yes!! That would be wonderful. Awesome that this is part of your project's scope! :tada: 

Everything I've done is above in https://github.com/cfpb/design-manual/issues/281#issuecomment-163691016. My team hasn't done any FEWD work on it. And I haven't started on the DM page but am happy to still handle that— Whatever's easier for your team. We can talk more during the DM regroup meeting.

**Comment by niqjohnson** [comment]
Just a note that one item in the original spec jumped out at me:

> Pages will print in full color, and all images will print

Browser print dialogs usually have a "print background colors and images" option that's usually unchecked by default. It's been a while since I've dug into print styles, but last I knew it took some CSS trickery to always print background colors and images exactly as they appeared on the screen (and there wasn't a global solution, only a bunch of situation-specific ones).

We'll do our best to stick to the original spec and reassess as we run into problems. Onward!

**Comment by KimberlyMunoz** [comment]
>  I knew it took some CSS trickery to always print background colors and images exactly as they appeared on the screen

This is like the decision on grayscale images, it's best to avoid overwriting intended printer behavior especially with anything non-standard. The user can check that box if they want the full color/image experience and we don't want to force people into wasting more money on ink if they intend to drop out background colors and images.

**Comment by gboone** [comment]
Hi friends! I'm working on reviewing a PR to resolve 18F/18f.gsa.gov#1606. I'm curious if someone would be willing to share any research or background on the decision to expand URLs in the print stylesheet? Our concern is that for a blog post, the expanded URLs might impede readability of the printed document. Particularly if there are many of them or when the expanded URLs are quite long. Thanks, and hello from 18F!

**Comment by jenn-franklin** [comment]
Hi @gboone! Here are two articles that may or may not be helpful to you: <a href="https://www.smashingmagazine.com/2013/03/tips-and-tricks-for-print-style-sheets/">Tips and Tricks For Print Style Sheets</a> and <a href="http://www.webdesignerdepot.com/2010/01/10-tips-for-better-print-style-sheets/">10 Tips for Better Print Style Sheets</a>. We felt that expanding URLs within the content specific to that page would be helpful in providing users with context and further information. We decided it was unnecessary to expand nav links. Certainly hear what you're saying regarding readability, which is one of the reasons we decided to differentiate the font weight between the link text and path. (<a href="https://github.com/cfpb/design-manual/issues/281#issuecomment-98703749">Here are some explorations of that</a>.) @marteki @niqjohnson @caheberer, certainly chime in if y'all have changed directions at all with this.

**Comment by stephanieosan** [comment]
@huetingj We need some status updates on this: 

What is the status of print style sheets on V1/Capital Framework? 
How does that align or not align with your proposed print style guidelines? 
What needs to be done to proceed on this issue?

**Comment by jenn-franklin** [comment]
@marteki @niqjohnson @caheberer I believe your team was going to dig into print styles while working on Dynamic Disclosures. Was anything contributed to CF, and does it align with the proposed guidelines above?

**Comment by caheberer** [comment]
I don't have any info on this...not sure where we left off development-wise, so Nicholas and Marteki should have a better idea.

**Comment by caheberer** [comment]
@niqjohnson @marteki any update on this?

**Comment by marteki** [comment]
We worked on this some back in late 2015 and early 2016, but it kept being de-prioritized in the project. When I look back at my archived code and branches for both `college-costs` and `capital-framework`, I can't find any of the relevant print styling code that I know I wrote.

I think it's been lost. 😢

**Comment by caheberer** [comment]
Thanks for the update. I will add it to our backlog.
